samples/order/processor/processor.go

```diff
@@ -0,0 +1,251 @@
+	"time"
+
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
+	"gocloud.dev/docstore"
+	_ "gocloud.dev/docstore/memdocstore"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/mempubsub"
+	"gocloud.dev/samples/order/internal/common"
+)
+
+var (
+	requestSubURL    = flag.String("request-sub", "", "URL for request pub/sub subscription")
+	responseTopicURL = flag.String("response-topic", "", "URL for response pub/sub topic")
+	bucketURL        = flag.String("bucket", "", "URL for image bucket")
+	collectionURL    = flag.String("collection", "", "URL for order collection")
+)
+
+func main() {
+	flag.Parse()
+	if *requestSubURL == "" {
+		log.Fatal("missing -request-sub")
+	}
+	if *responseTopicURL == "" {
+		log.Fatal("missing -response-topic")
+	}
+	if *bucketURL == "" {
+		log.Fatal("missing -bucket")
+	}
+	if *collectionURL == "" {
+		log.Fatal("missing -collection")
+	}
+	ctx := context.Background()
+	p, err := newProcessor(ctx, *requestSubURL, *responseTopicURL, *bucketURL, *collectionURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if err := p.run(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+type processor struct {
+	requestSub    *pubsub.Subscription
+	responseTopic *pubsub.Topic
+	bucket        *blob.Bucket
+	coll          *docstore.Collection
+}
+
+func newProcessor(ctx context.Context, subURL, topicURL, bucketURL, collURL string) (*processor, error) {
+	reqSub, err := pubsub.OpenSubscription(ctx, subURL)
+	if err != nil {
+		return nil, err
+	}
+	resTopic, err := pubsub.OpenTopic(ctx, topicURL)
+	if err != nil {
+		return nil, err
+	}
+	bucket, err := blob.OpenBucket(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+	coll, err := docstore.OpenCollection(ctx, collURL)
+	if err != nil {
+		return nil, err
+	}
+	return &processor{
+		requestSub:    reqSub,
+		responseTopic: resTopic,
+		bucket:        bucket,
+		coll:          coll,
+	}, nil
+}
+
+func (p *processor) run(ctx context.Context) error {
+	for {
+		if err := p.handleRequest(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *processor) handleRequest(ctx context.Context) error {
+	var req common.OrderRequest
+	msg, err := receiveJSON(ctx, p.requestSub, &req)
+	// Ack the message because we handled it, even on error.
+	defer func() {
+		if msg != nil {
+			msg.Ack()
+		}
+	}()
+	if err != nil {
+		return err
+	}
+	log.Printf("received %+v", req)
+	res := p.handleOrder(ctx, &req)
+	if res == nil {
+		log.Printf("duplicate finished order %v", req.ID)
+		return nil
+	}
+	if err := sendJSON(ctx, p.responseTopic, res); err != nil {
+		// TODO(jba): should we really ack in this case? Not clear.
+		return err
+	}
+	log.Printf("sent %+v", res)
+	return nil
+}
+
+// handleOrder processes the order request. A processing error is a kind of response.
+func (p *processor) handleOrder(ctx context.Context, req *common.OrderRequest) *common.OrderResponse {
+	res, err := p.processOrder(ctx, req)
+	if err != nil {
+		// TODO(jba): record error metric
```

Comment:
What is the gain for having handleOrder wrap processOrder, if it only handles the error? processOrder already has a defer to handle errors, couldn't the wrapping with OrderResponse just happen there?

Reply:
The code would be odd and error-prone because it would need to communicate the error to the defer. Something like:
```
var err error
defer func() { if err != nil { ...} }()
...
if err2 := p.coll.Create(ctx, order); err2 != nil {
    err = err2
    return nil
}
```
I think it will be harder to understand that the existing code.

Perhaps I'm missing a more elegant solution; please sketch it for me.